src/libopenrave-core/jsonparser/jsonreader.cpp

```diff
@@ -298,6 +298,36 @@ class JSONReader
                 }
                 return false;
             }
+            // if parent rEnvInfo is going to overwrite bodies in the environment of the referenceUri, do not deserialize them. otherwise, they would be merged instead of overwritten.
+            if( rEnvInfo.HasMember("bodies") && rEnvInfo["bodies"].Size() > 0 ) {
+                if( prReferenceEnvInfo->HasMember("bodies") && (*prReferenceEnvInfo)["bodies"].Size() > 0 ) {
+                    boost::shared_ptr<rapidjson::Document> prFilteredReferenceEnvInfo = boost::make_shared<rapidjson::Document>(&alloc);
+                    prFilteredReferenceEnvInfo->CopyFrom(*prReferenceEnvInfo, alloc);
```

Comment:
Might be nice if we could avoid copies, though I guess reference environments should generally be smaller? Erasing the elements from the original list seems slow (rapidjson `Erase` has to `memmove` all subsequent), but maybe just `SetNull` on the items to filter and skip non-objects when processing?

Reply:
We need to keep `prReferenceEnvInfo` untouched since it is cached. We can think of making shallow copy of a rapidjson object to avoid deepcopy for speed up but I think it would be a little big hustle and I am not sure if it is worth an effort as this procedure currently happens only on startup and reference environments should be compact compared to dynamic scene state with mobile robots as you mentioned.
`SetNull` instead of `Erase` requires many other places to be changed. I guess we don't want that only for this also.